File: source_py/ebook_renamer/duplicates.py
"""
Duplicate detection functionality for the ebook renamer.
"""

import hashlib
import os
from typing import List, Tuple

from .types import FileInfo


class DuplicateDetector:
    """Handles duplicate detection based on MD5 hash."""
    
    # Allowed formats to keep
    ALLOWED_EXTENSIONS = {".pdf", ".epub", ".txt"}
    
    def detect_duplicates(self, files: List[FileInfo]) -> Tuple[List[List[str]], List[FileInfo]]:
        """Find duplicate files based on MD5 hash."""
        # Filter to only allowed formats first
        filtered_files = [file for file in files 
                         if file.extension in self.ALLOWED_EXTENSIONS]
        
        # Build hash map: file_hash -> list of file infos
        hash_map = {}
        
        for file_info in filtered_files:
            if not file_info.is_failed_download and not file_info.is_too_small:
                try:
                    file_hash = self._compute_md5(file_info.original_path)
                    if file_hash not in hash_map:
                        hash_map[file_hash] = []
                    hash_map[file_hash].append(file_info)
                except (OSError, IOError):
                    # Skip files that can't be read
                    continue
        
        # Group duplicates by hash and apply retention strategy
        duplicate_groups = []
        duplicate_paths = set()
        
        for file_infos in hash_map.values():
            if len(file_infos) > 1:
                # Multiple files with same hash - apply retention strategy
                kept_file = self._select_file_to_keep(file_infos)
                
                group_paths = [kept_file.original_path]
                
                for file_info in file_infos:
                    if file_info.original_path != kept_file.original_path:
                        duplicate_paths.add(file_info.original_path)
                        group_paths.append(file_info.original_path)
                
                duplicate_groups.append(group_paths)
        
        # Return only non-duplicate files (including filtered out formats)
        clean_files = [file for file in filtered_files 
                      if file.original_path not in duplicate_paths]
        
        return duplicate_groups, clean_files
    
    def _select_file_to_keep(self, files: List[FileInfo]) -> FileInfo:
        """Select the file to keep based on priority: normalized > shortest path > newest."""
        # Priority 1: Already normalized files (have new_name set)
        normalized_files = [file for file in files if file.new_name is not None]
        candidates = normalized_files if normalized_files else files
        
        # Priority 2: Shortest path (fewest directory components) among candidates
        candidates_with_depth = []
        min_depth = float('inf')
        
        for file in candidates:
            depth = file.original_path.count('/')
            candidates_with_depth.append((depth, file))
            if depth < min_depth:
                min_depth = depth
        
        # Filter to shallowest candidates
        shallowest_candidates = [file for depth, file in candidates_with_depth 
                                if depth == min_depth]
        
        # Priority 3: Newest modification time among the shallowest candidates
        if not shallowest_candidates:
            # Fallback: return first file
            return files[0]
        
        newest_file = shallowest_candidates[0]
        for file in shallowest_candidates[1:]:
            if file.modified_time > newest_file.modified_time:
                newest_file = file
        
        return newest_file
    
    def _compute_md5(self, file_path: str) -> str:
        """Calculate MD5 hash of a file."""
        hash_md5 = hashlib.md5()
        with open(file_path, "rb") as f:
            for chunk in iter(lambda: f.read(4096), b""):
                hash_md5.update(chunk)
        return hash_md5.hexdigest()
    
    def detect_name_variants(self, files: List[FileInfo]) -> List[List[int]]:
        """Group files by normalized name (treating (1), (2), etc. as variants)."""
        # Group files by normalized name (treating (1), (2), etc. as variants)
        name_groups = {}
        
        for idx, file_info in enumerate(files):
            if file_info.new_name is not None:
                # Strip off (1), (2), etc. to find base name
                base_name = self._strip_variant_suffix(file_info.new_name)
                if base_name not in name_groups:
                    name_groups[base_name] = []
                name_groups[base_name].append(idx)
        
        # Keep only groups with duplicates
        variants = [group for group in name_groups.values() if len(group) > 1]
        return variants
    
    def _strip_variant_suffix(self, filename: str) -> str:
        """Strip patterns like " (1)", " (2)", etc. from the end before extension."""
        # Match patterns like " (1)", " (2)", etc. at the end before extension
        if '.' in filename:
            name_part = filename.rsplit('.', 1)[0]
            ext_part = filename.rsplit('.', 1)[1]
            
            # Remove variant suffix from name part
            if name_part.endswith(')'):
                # Check if it matches pattern " (n)"
                open_paren = name_part.rfind(' (')
                if open_paren != -1:
                    suffix = name_part[open_paren:]
                    if len(suffix) >= 4 and suffix.startswith(' (') and suffix.endswith(')'):
                        # Check if content between parens is numeric
                        content = suffix[2:-1]
                        if content.isdigit():
                            name_part = name_part[:open_paren]
            
            return f"{name_part}.{ext_part}"
        else:
            # No extension, just check for variant suffix
            if filename.endswith(')'):
                open_paren = filename.rfind(' (')
                if open_paren != -1:
                    suffix = filename[open_paren:]
                    if len(suffix) >= 4 and suffix.startswith(' (') and suffix.endswith(')'):
                        content = suffix[2:-1]
                        if content.isdigit():
                            return filename[:open_paren]
        
        return filename

File: source_go/internal/duplicates/duplicates.go
package duplicates

import (
	"crypto/md5"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"

	"github.com/ebook-renamer/go/internal/types"
)

// Allowed formats to keep
var allowedExtensions = map[string]bool{
	".pdf": true,
	".epub": true,
	".txt": true,
}

// DetectDuplicates finds duplicate files based on MD5 hash
func DetectDuplicates(files []types.FileInfo) ([][]string, []types.FileInfo, error) {
	// Filter to only allowed formats first
	var filteredFiles []types.FileInfo
	for _, file := range files {
		if allowedExtensions[file.Extension] {
			filteredFiles = append(filteredFiles, file)
		}
	}

	// Build hash map: file_hash -> list of file infos
	hashMap := make(map[string][]types.FileInfo)

	for _, fileInfo := range filteredFiles {
		if !fileInfo.IsFailedDownload && !fileInfo.IsTooSmall {
			hash, err := computeMD5(fileInfo.OriginalPath)
			if err != nil {
				// Skip files that can't be read
				continue
			}
			hashMap[hash] = append(hashMap[hash], fileInfo)
		}
	}

	// Group duplicates by hash and apply retention strategy
	var duplicateGroups [][]string
	duplicatePaths := make(map[string]bool)

	for _, fileInfos := range hashMap {
		if len(fileInfos) > 1 {
			// Multiple files with same hash - apply retention strategy
			keptFile := selectFileToKeep(fileInfos)
			
			var groupPaths []string
			groupPaths = append(groupPaths, keptFile.OriginalPath)
			
			for _, fileInfo := range fileInfos {
				if fileInfo.OriginalPath != keptFile.OriginalPath {
					duplicatePaths[fileInfo.OriginalPath] = true
					groupPaths = append(groupPaths, fileInfo.OriginalPath)
				}
			}

			duplicateGroups = append(duplicateGroups, groupPaths)
		}
	}

	// Return only non-duplicate files (including filtered out formats)
	var cleanFiles []types.FileInfo
	for _, file := range filteredFiles {
		if !duplicatePaths[file.OriginalPath] {
			cleanFiles = append(cleanFiles, file)
		}
	}

	return duplicateGroups, cleanFiles, nil
}

// selectFileToKeep selects the file to keep based on priority: normalized > shortest path > newest
func selectFileToKeep(files []types.FileInfo) types.FileInfo {
	// Priority 1: Already normalized files (have new_name set)
	var normalizedFiles []types.FileInfo
	var allFiles []types.FileInfo
	
	for _, file := range files {
		allFiles = append(allFiles, file)
		if file.NewName != nil {
			normalizedFiles = append(normalizedFiles, file)
		}
	}

	// Use normalized files if available, otherwise use all files
	candidates := normalizedFiles
	if len(candidates) == 0 {
		candidates = allFiles
	}

	// Priority 2: Shortest path (fewest directory components) among candidates
	type depthFile struct {
		depth int
		file  types.FileInfo
	}
	
	var candidatesWithDepth []depthFile
	minDepth := int(^uint(0) >> 1) // Max int
	
	for _, file := range candidates {
		depth := strings.Count(file.OriginalPath, string(filepath.Separator))
		candidatesWithDepth = append(candidatesWithDepth, depthFile{depth, file})
		if depth < minDepth {
			minDepth = depth
		}
	}
	
	// Filter to shallowest candidates
	var shallowestCandidates []types.FileInfo
	for _, df := range candidatesWithDepth {
		if df.depth == minDepth {
			shallowestCandidates = append(shallowestCandidates, df.file)
		}
	}

	// Priority 3: Newest modification time among the shallowest candidates
	if len(shallowestCandidates) == 0 {
		// Fallback: return first file
		return files[0]
	}

	newestFile := shallowestCandidates[0]
	for _, file := range shallowestCandidates[1:] {
		if file.ModifiedTime.After(newestFile.ModifiedTime) {
			newestFile = file
		}
	}

	return newestFile
}

// computeMD5 calculates MD5 hash of a file
func computeMD5(filePath string) (string, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return "", err
	}
	defer file.Close()

	hash := md5.New()
	if _, err := io.Copy(hash, file); err != nil {
		return "", err
	}

	return fmt.Sprintf("%x", hash.Sum(nil)), nil
}

// DetectNameVariants groups files by normalized name (treating (1), (2), etc. as variants)
func DetectNameVariants(files []types.FileInfo) [][]int {
	// Group files by normalized name (treating (1), (2), etc. as variants)
	nameGroups := make(map[string][]int)

	for idx, fileInfo := range files {
		if fileInfo.NewName != nil {
			// Strip off (1), (2), etc. to find base name
			baseName := stripVariantSuffix(*fileInfo.NewName)
			nameGroups[baseName] = append(nameGroups[baseName], idx)
		}
	}

	// Keep only groups with duplicates
	var variants [][]int
	for _, group := range nameGroups {
		if len(group) > 1 {
			variants = append(variants, group)
		}
	}

	return variants
}

// stripVariantSuffix strips patterns like " (1)", " (2)", etc. from the end before extension
func stripVariantSuffix(filename string) string {
	// Match patterns like " (1)", " (2)", etc. at the end before extension
	if dotIdx := strings.LastIndex(filename, "."); dotIdx != -1 {
		namePart := filename[:dotIdx]
		extPart := filename[dotIdx:]
		
		// Remove variant suffix from name part
		if strings.HasSuffix(namePart, ")") {
			// Check if it matches pattern " (n)"
			openParen := strings.LastIndex(namePart, " (")
			if openParen != -1 {
				suffix := namePart[openParen:]
				if len(suffix) >= 4 && suffix[0] == ' ' && suffix[1] == '(' {
					// Check if content between parens is numeric
					content := suffix[2 : len(suffix)-1]
					isNumeric := true
					for _, r := range content {
						if r < '0' || r > '9' {
							isNumeric = false
							break
						}
					}
					if isNumeric {
						namePart = namePart[:openParen]
					}
				}
			}
		}
		
		return namePart + extPart
	} else {
		// No extension, just check for variant suffix
		if strings.HasSuffix(filename, ")") {
			openParen := strings.LastIndex(filename, " (")
			if openParen != -1 {
				suffix := filename[openParen:]
				if len(suffix) >= 4 && suffix[0] == ' ' && suffix[1] == '(' {
					content := suffix[2 : len(suffix)-1]
					isNumeric := true
					for _, r := range content {
						if r < '0' || r > '9' {
							isNumeric = false
							break
						}
					}
					if isNumeric {
						return filename[:openParen]
					}
				}
			}
		}
	}

	return filename
}
